First commit of CILogon authentication

skoranda opened the pull request:
First commit of CILogon authentication. CILogon provides a
standards-compliant OpenID Connect (OAuth 2.0) interface to federated
authentication including InCommon, the Australian Access Federation
(AAF), and eduGAIN. CILogon OpenID Connect (OIDC) client registration
is available to researchers and scholars at
https://cilogon.org/oauth2/register

<!--
Make sure to follow the DEV guidelines (https://gen3.org/resources/developer/dev-introduction/) before asking for review.

- Describe what this pull request does.
- Add short descriptive bullet points for each section if relevant. Keep in mind that they will be parsed automatically to generate official release notes.
- Test manually.
- Maintain or increase the test coverage (if relevant).
- Update the documentation, or justify if not needed.

-->

### New Features
Add CILogon as an authentication option. CILogon provides a standards-compliant OpenID Connect (OAuth 2.0) interface to federated authentication including InCommon, the Australian Access Federation (AAF), and eduGAIN. CILogon OpenID Connect (OIDC) client registration is available to researchers and scholars at https://cilogon.org/oauth2/register

ohofmann commented:
Thanks Scott! For context, some of the upcoming Gen3 instances in Australia are working with the Australian Access Federation to support a unified user authentication approach, and CILogon is part of this pilot.

xritter1 commented:
Hi! Looks good! Could you please update the base branch and remove this variable "state" [line](https://github.com/uc-cdis/fence/pull/896/files#diff-539c11773050a2fccd80045eb962a38e3bf6424677dd529d612a1733ccd385a5R31-R33)?

paulineribeyre commented on fence/blueprints/login/cilogon.py:
```diff
@@ -0,0 +1,20 @@
+import flask
+
+from fence.resources.openid.cilogon_oauth2 import CILOGON_IDP_NAME
```
I would declare `CILOGON_IDP_NAME` in this file instead of `fence/resources/openid/cilogon_oauth2.py`, since it's only used in this file

paulineribeyre reviewed (commented):
Looks great, thanks for the contribution! I just have a small comment below - and we'll test this before approving 🙂

paulineribeyre commented:
Tested by @williamhaley ✔️ 
@skoranda, we can merge once the branch is up to date

skoranda commented:
@paulineribeyre I implemented your request regarding the variable CILOGON_IDP_NAME, rebased on the current HEAD of main, reran black, and verified that the test suite passed all tests. Please let me know if you require any additional changes. Thanks.
